server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java

```diff
@@ -110,4 +153,73 @@ public static QueryExplanation readQueryExplanation(StreamInput in)  throws IOEx
         exp.readFrom(in);
         return exp;
     }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (getIndex() != null) {
+            builder.field(INDEX_FIELD, getIndex());
+        }
+        if(getShard() >= 0) {
+            builder.field(SHARD_FIELD, getShard());
+        }
+        builder.field(VALID_FIELD, isValid());
+        if (getError() != null) {
+            builder.field(ERROR_FIELD, getError());
+        }
+        if (getExplanation() != null) {
+            builder.field(EXPLANATION_FIELD, getExplanation());
+        }
+        return builder;
+    }
+
+    public static QueryExplanation fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof QueryExplanation) {
+            QueryExplanation other = (QueryExplanation) o;
+            boolean result;
+            if (getIndex() == null) {
+                result = other.getIndex() == null;
+            } else {
+                result = getIndex().equals(other.getIndex());
+            }
+            result &= this.getShard() == other.getShard();
+            result &= isValid() == other.isValid();
+            if (getError() == null) {
+                result &= other.getError() == null;
+            } else {
+                result &= getError().equals(other.getError());
+            }
+            if (getExplanation() == null) {
+                result &= other.getExplanation() == null;
+            } else {
+                result &= getExplanation().equals(other.getExplanation());
+            }
+            return result;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        if (getIndex() != null) {
```

Comment:
I'd use `Objects.hash` because the performance here isn't that important and it is easier to review for correctness.

Reply:
Thanks! I didn't know this existed either :)